### PR TITLE
add feature type override in signals

### PR DESCRIPTION
**Pull request by ycheng35xo**
1 add feature type override in signal ymal
2 add common util function for is_categorical() and is_numerical()

### CODE REVIEW

File: assets/model_monitoring/components/data_drift/data_drift_compute_metrics/spec.yaml [hunk 1]
```diff
@@ -29,6 +29,12 @@ inputs:
     type: number
   numerical_threshold:
     type: number
+  override_numerical_feature:
+    type: string
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/data_drift/data_drift_compute_metrics/spec.yaml:33]: Is it a formatted string that we can parse multiple feature override entries from it?

File: assets/model_monitoring/components/src/data_drift_compute_metrics/run.py [hunk 2]
```diff
@@ -23,6 +23,8 @@ def run():
     parser.add_argument("--numerical_metric", type=str)
     parser.add_argument("--categorical_threshold", type=str)
     parser.add_argument("--numerical_threshold", type=str)
+    parser.add_argument("--override_numerical_feature", type=str)
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/data_drift_compute_metrics/run.py:26]: can we add nargs='+' to make it a list of str?

**Comment by VivienneTang** [comment]
Why do we only override a feature instead of a list of features?

File: assets/model_monitoring/components/data_drift/data_drift_compute_metrics/spec.yaml (hunk 1 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/data_drift/data_drift_compute_metrics/spec.yaml:33]: yes I just followed the pattern of this categorical_metrics. It is a comma-joint string. For example, if user wants to override UserId and TransactionId as numerical value. This will be "UserId,TransactionId".

https://github.com/Azure/azureml-assets/blob/4667dcbbda9316906a718c85b7f878873b3523b2/assets/model_monitoring/components/data_quality/data_quality_compute_metrics/spec.yaml#L22C7-L22C7

File: assets/model_monitoring/components/src/data_drift_compute_metrics/run.py (hunk 2 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/data_drift_compute_metrics/run.py:26]: It is a comma-joint string so we can just do a split(",")

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py [hunk 3]
```diff
@@ -35,6 +35,8 @@ def parse_args():
     parser.add_argument("--target_column", type=str, required=False)
     parser.add_argument("--task_type", type=str)
     parser.add_argument("--signal_metrics", type=str)
+    parser.add_argument("--override_numerical_features", type=str)
+    parser.add_argument("--override_categorical_features", type=str)
```
Comment (VivienneTang) [inline assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:39]: do you need to put " required=False" as they are optional?

File: assets/model_monitoring/components/data_drift/data_drift_compute_metrics/spec.yaml [hunk 4]
```diff
@@ -72,4 +78,6 @@ args: >-
   --numerical_threshold ${{inputs.numerical_threshold}}
   --categorical_threshold ${{inputs.categorical_threshold}}
   --feature_names ${{inputs.feature_names}}
+  --override_numerical_features ${{inputs.override_numerical_features}}
```
Comment (VivienneTang) [inline assets/model_monitoring/components/data_drift/data_drift_compute_metrics/spec.yaml:81]: Do you need to put [[]] for the optional args?

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py [hunk 5]
```diff
@@ -268,10 +271,14 @@ def run(args):
         log_time_and_message("Reading data in spark and converting to pandas")
         baseline_df = try_read_mltable_in_spark_with_error(args.baseline_data, "baseline_data")
 
+        raw_categorical_features = get_categorical_cols_with_df_with_override(baseline_df,
```
Comment (hawestra) [inline assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:274]: this variable isn't used anywhere else. if this is meant to be used once the comment on line 281 is addressed, can we comment this line out and add a TODO here too. Otherwise, it might unnecessarily introduce issues and is confusing

File: assets/model_monitoring/components/feature_attribution_drift/feature_importance_metrics/spec.yaml [hunk 6]
```diff
@@ -21,6 +21,12 @@ inputs:
   task_type:
     type: string
     optional: true
+  override_numerical_features:
```
Comment (hawestra) [inline assets/model_monitoring/components/feature_attribution_drift/feature_importance_metrics/spec.yaml:24]: would be great if we could add some UI warning that says that adjusting these might cause issues with computing feature importance if the columns can't be properly cast; obviously not part of this PR just something to note

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py [hunk 7]
```diff
@@ -34,6 +38,98 @@ def get_categorical_columns(column_dtype_map: dict) -> list:
...
+        return False
+    if column_dtype_map[column] in data_type_categorical_group:
+        return True
+    if column_dtype_map[column] in data_type_numerical_group:
+        return False
+    if column_dtype_map[column] in data_type_long_group:
+        distinct_value_ratio = get_distinct_ratio(df.select(column).rdd.flatMap(lambda x: x).collect())
+        return distinct_value_ratio < 0.05
+
+    print(f"Unknown column type: {column_dtype_map[column]}, column name: {column}")
```
Comment (hawestra) [inline assets/model_monitoring/components/src/shared_utilities/df_utils.py:73]: you may want to log this, so you could have telemetry on which datatypes you're missing

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py [hunk 8]
```diff
@@ -270,10 +271,14 @@ def run(args):
...
+                                                   override_numerical_features=args.override_numerical_features,
+                                                   override_categorical_features=args.override_categorical_features)
         baseline_df = baseline_df.toPandas()
         task_type = determine_task_type(args.task_type, args.target_column, baseline_df)
         log_time_and_message(f"Computed task type is {task_type}")
 
+        # Todo: get this from raw_categorical_features (removing target column, validate feature types)
         categorical_features = compute_categorical_features(baseline_df, args.target_column)
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:282]: Is this compute_categorical_features() written by Shreeya with some complex logic? I suppose with feature_override, we'll replace this with get_categorical_cols_with_df_with_df_with_override()?

File: assets/model_monitoring/components/src/model_monitor_compute_histogram/histogram.py [hunk 9]
```diff
@@ -130,19 +130,17 @@ def _to_bin_edges(histogram_buckets: pyspark_sql.DataFrame):
...
-    df_dtypes = dict(df.dtypes)
-    for (column_name, data_type) in df_dtypes.items():
-        columns_dict[column_name] = data_type
-
-    numerical_columns = get_numerical_cols_with_df(columns_dict,
-                                                   df)
-    categorical_columns = get_categorical_cols_with_df(columns_dict,
-                                                       df)
+    numerical_columns_names, categorical_columns_names = get_numerical_and_categorical_cols(
+                                                            baseline_df,
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/model_monitor_compute_histogram/histogram.py:140]: where comes this baseline_df? And the below common_columns_dict?

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py [hunk 10]
```diff
@@ -34,6 +38,98 @@ def get_categorical_columns(column_dtype_map: dict) -> list:
...
+    if feature_type_override_map.get(column, None) == "numerical":
+        return True
+    if feature_type_override_map.get(column, None) == "categorical":
+        return False
+    if column_dtype_map[column] in data_type_numerical_group:
+        return True
+    if column_dtype_map[column] in data_type_categorical_group:
+        return False
+    if column_dtype_map[column] in data_type_long_group:
+        distinct_value_ratio = get_distinct_ratio(df.select(column).rdd.flatMap(lambda x: x).collect())
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/shared_utilities/df_utils.py:52]: with override feature, do we still want to apply this logic to determine numerical/categorical column?

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py [hunk 11]
```diff
@@ -6,8 +6,11 @@
 
 data_type_double_group = ["float", "double"]
 data_type_long_group = ["long", "int", "bigint", "short"]
+data_type_numerical_group = ["float", "double", "decimal"]
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/shared_utilities/df_utils.py:9]: can we merge data_type_double_group and data_type_numerical_group?

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py [hunk 12]
```diff
@@ -6,8 +6,11 @@
...
+data_type_categorical_group = ["string", "boolean", "timestamp", "date", "binary"]
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/shared_utilities/df_utils.py:10]: why we want to treat timestamp, binary as categorical columns by default? Any guideline?

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py [hunk 13]
```diff
@@ -34,6 +38,98 @@ def get_categorical_columns(column_dtype_map: dict) -> list:
...
+        return False
+    if column_dtype_map[column] in data_type_long_group:
+        distinct_value_ratio = get_distinct_ratio(df.select(column).rdd.flatMap(lambda x: x).collect())
+        return distinct_value_ratio >= 0.05
+
+    print(f"Unknown column type: {column_dtype_map[column]}, column name: {column}")
+    return False
+
+
+def is_categorical_new(column, column_dtype_map: dict, feature_type_override_map: dict, df):
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/shared_utilities/df_utils.py:59]: seems we can define is_categorical_new as not is_numerical_new()?

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py [hunk 14]
```diff
@@ -34,6 +38,98 @@ def get_categorical_columns(column_dtype_map: dict) -> list:
...
+        override_categorical_features,
+        column_dtype_map=None,):
+    return (get_numerical_cols_with_df_with_override(df, column_dtype_map, override_numerical_features, override_categorical_features),
+            get_categorical_cols_with_df_with_override(df, column_dtype_map, override_numerical_features, override_categorical_features))
+
+
+def get_feature_type_override_map(override_numerical_features: str, override_categorical_features: str) -> dict:
+    """ Generate feature type override map with key of feature name and value of "numerical"/"categorical"."""
+    feature_type_override_map = {}
+    if override_categorical_features is not None:
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/shared_utilities/df_utils.py:123]: can simplified as
```py
if override_categorical_features:
    ...
```

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py [hunk 15]
```diff
@@ -34,6 +38,98 @@ def get_categorical_columns(column_dtype_map: dict) -> list:
...
+
+    print(f"Unknown column type: {column_dtype_map[column]}, column name: {column}")
+    return False
+
+
+def get_numerical_cols_with_df_with_override(
+        df,
+        override_numerical_features,
+        override_categorical_features,
+        column_dtype_map=None,) -> list:
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/shared_utilities/df_utils.py:81]: is this column_dtype_map just for easy UT? if so, you can do provide a mock df with mock df.dtypes for UT.

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 10 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py:52]: I think so, because override feature is optional. If the user did not specify, it would be good to keep it

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 15 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py:81]: no in the data drift component, the logic is to only get the features in the common columns, and this is a required parameter for data drift component

File: assets/model_monitoring/components/src/model_monitor_compute_histogram/histogram.py (hunk 9 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/model_monitor_compute_histogram/histogram.py:140]: sorry I did not update this file. It is the old version.

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 12 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py:10]: I checked with PM. According to PM: 

"It depends a bit on the customer scenario. For timestamp, we can default to categorical because it may shift a lot, if it is smooth it may make more sense for it to be numerical. For binary it’s by definition categorical"

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 13 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py:59]: technically yes. But there might be some rare cases that we have some unknow type columns, now we want to log them as mark them neither numerical nor categorical. I will add a todo and if we do not see so many unknow types, we can default to categorical maybe

`    print(f"Unknown column type: {column_dtype_map[column]}, column name: {column}")`

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py (hunk 5 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:276]: updated the change

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py (hunk 8 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py]: done the changes

File: assets/model_monitoring/components/src/data_drift_compute_metrics/run.py (hunk 2 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/data_drift_compute_metrics/run.py]: then please rename the parameter to --override_numerical_feature**s**

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py [hunk 16]
```diff
@@ -270,17 +273,25 @@ def run(args):
         log_time_and_message("Reading data in spark and converting to pandas")
         baseline_df = try_read_mltable_in_spark_with_error(args.baseline_data, "baseline_data")
 
+        categorical_features = get_categorical_cols_with_df_with_override(baseline_df,
+                                                                              args.override_numerical_features,
+                                                                              args.override_categorical_features)
         baseline_df = baseline_df.toPandas()
-        task_type = determine_task_type(args.task_type, args.target_column, baseline_df)
+        # lightgbm does not support features that can't be converted to non-bool, int, float features
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:280]: what's non-bool type?
In general I don't quite understand this comment. Do you mean lightgbm only supports features that can be converted to bool, int, float? So we have to filter out more categorical features returned from common logic?

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py [hunk 17]
```diff
@@ -270,17 +273,25 @@ def run(args):
...
+        # if these features can't be converted to an int. Therefore we have to more these as "category"
+        # types
+        categorical_features_lgbm = compute_lightgbm_unsupported_categorical_features(baseline_df,
```
Comment (RichardLi1437) [inline assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:283]: I'm confused with the naming here, is categroical_features_lgbm stores for supported or unsupported lgbm categorical features?
Does this method do some lightgbm specific logic to further filter out some categorical columns? If so, maybe rename to sth like `get_lightgbm_categrocial_features()`

Comment (RichardLi1437) [inline assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:283]: also, I suppose this compute_feature_importance component is not lightgbm specific? why apply lightgbm speicifc logic here?

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 10 above)
Comment (RichardLi1437) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py:51]: this logic is complicated and yet not guarantee it is what user want, IMO, if no feature override, maybe just simply determine the numerical/categorical by dtype. You can discuss with Yunjie and Feng to see if we want to keep it.

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py (hunk 17 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py:283]: According to Hannah, this is the current limitation. Now we only use the lightgbm model for the feature importance calculation. Ideally we'd use the model that the user is using

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 10 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py:51]: but for the integer type, we cannot know if it is numerical or categorical. I think that is the common design for determining the integer type?

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py (hunk 16 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py]: I modified the comment. From my understanding:
    # Lightgbm can only support features that can be converted to bool, int, float.
    # If these features can't be converted, we have to mark them as "category" types so lightgbm will ignore them.
    # In our design, we mark all known categorical features (including bool) as "category", 
    # only filter out datetime and timedelta because they can be converted to int

I will ask Hannah to review the feature importance part

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py (hunk 17 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py]: Done

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 13 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py]: return None for unknown

File: assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py [hunk 18]
```diff
@@ -56,47 +54,47 @@ def get_df_schema(df: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
...
         The DataFrame has two columns:
         featureName: The name of the categorical column.
         set: A list of unique values for the categorical column.
     """
-    # Define a list of categorical column names to process
-    cat_col_names = [c[0] for c in df.dtypes if c[1] == "string"]
+    # Ignore bool, time, date categorical columns because they are meaningless for data quality calculation
+    modified_categorical_columns = []
+    dtype_map = dict(df.dtypes)
+    for column in categorical_columns:
```
Comment (VivienneTang) [inline assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py:90]: please add unit test. We used to get unique value only for string. With your change, we are get unique value for more datatype.

**Review by hawestra — APPROVED** [review]

File: assets/model_monitoring/components/tests/unit/test_compute_data_quality_statistics.py [hunk 19]
```diff
@@ -129,7 +129,6 @@ def test_compute_data_quality_statistics(
     ):
         """Test compute data quality statistics with string, integer, boolean, double type."""
         actual_data_stats_table = compute_data_quality_statistics(df_with_timestamp, None, None)
-        assert data_stats_table.count() == actual_data_stats_table.count()
```
Comment (VivienneTang) [inline assets/model_monitoring/components/tests/unit/test_compute_data_quality_statistics.py:104]: why do you delete this check?

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py [hunk 20]
```diff
@@ -161,30 +162,31 @@ def compute_max_violation(
...
 
     Returns:
         min_violation_df: A PySpark DataFrame with columns for violation count, feature name and metric name.
     """
     min_threshold_violation_count = []
     feature_name_list = []
 
-    numerical_types = ["double", "float", "int", "bigint", "smallint", "tinyint", "long"]
-
     for row in data_stats_table.filter(col("max_value").isNotNull()).select("featureName").distinct().collect():
```
Comment (VivienneTang) [inline assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py:183]: it is not your code, but could you change max_value to min_value?

File: assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py [hunk 21]
```diff
@@ -104,30 +104,31 @@ def get_null_count(df: pyspark.sql.DataFrame) -> pyspark.sql.DataFrame:
...
 
     Returns:
         max_violation_df: A PySpark DataFrame with columns for violation count, feature name and metric name.
     """
     max_threshold_violation_count = []
     feature_name_list = []
 
-    numerical_types = ["double", "float", "int", "bigint", "smallint", "tinyint", "long"]
-
     for row in data_stats_table.filter(col("min_value").isNotNull()).select("featureName").distinct().collect():
```
Comment (VivienneTang) [inline assets/model_monitoring/components/src/data_quality_compute_metrics/compute_data_quality_metrics.py:125]: it is not your code, but could you change min_value to max_value?

File: assets/model_monitoring/components/tests/unit/test_compute_data_quality_statistics.py (hunk 19 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/tests/unit/test_compute_data_quality_statistics.py:104]: misdeleted. Will add back

File: assets/model_monitoring/components/tests/unit/test_compute_data_quality_metrics.py [hunk 22]
```diff
@@ -205,3 +212,84 @@ def test_get_null_count(self):
...
+        df_with_missing_value = create_pyspark_dataframe(df_with_missing_value, schema)
+        df_with_inputed_value = create_pyspark_dataframe(df_with_inputed_value, schema)
+
+        numerical_columns = ["feature_double", "feature_integer"]
+        categorical_columns = ["feature_string"]
+        df_inputed_numerical = impute_numericals_with_median(df_with_missing_value, numerical_columns)
+        df_inputed_categorical = impute_categorical_with_mode(df_inputed_numerical, categorical_columns)
+
+        df_inputed_categorical.show()
+        print("final")
```
Comment (VivienneTang) [inline assets/model_monitoring/components/tests/unit/test_compute_data_quality_metrics.py:247]: clean up this line and the next line

File: assets/model_monitoring/components/tests/unit/test_compute_data_quality_metrics.py [hunk 23]
```diff
@@ -205,3 +212,84 @@ def test_get_null_count(self):
...
+            [
+                StructField("featureName", StringType(), True),
+                StructField("violationCount", IntegerType(), True),
+                StructField("metricName", StringType(), True),
+            ]
+        )
+        expected_set_violation_table = create_pyspark_dataframe(expected_set_violation_value, data_schema)
+
+        set_violation_table = compute_set_violation(df_input, data_stats_table, ["feature_string", "feature_char"])
+        set_violation_table.show()
```
Comment (VivienneTang) [inline assets/model_monitoring/components/tests/unit/test_compute_data_quality_metrics.py:293]: clean up?

File: assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py (hunk 18 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/data_quality_statistics/compute_data_quality_statistics.py]: now changed to only string

File: assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py (hunk 16 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/feature_importance_metrics/compute_feature_importance.py]: Now changed to all categorical features will be sent to lightgbm, without filtering out.

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 13 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py]: Fixed the change by making is_numerical() = Not is_categorical() and None for unknown type

File: assets/model_monitoring/components/src/shared_utilities/df_utils.py (hunk 10 above)
Comment (ycheng35xo) [reply assets/model_monitoring/components/src/shared_utilities/df_utils.py]: Checked with Yunjie. Keep the current implementation

**Review by VivienneTang — DISMISSED** [review]

**Review by VivienneTang — APPROVED** [review]
